### PR TITLE
support row constructor in IN expression 

### PR DESCRIPTION
**Pull request by siddontang**
e.g `select row(1, 1) in (row(1,1), row(1,0))`

### CODE REVIEW

File: expression/expressions/helper.go [hunk 1]
```diff
@@ -498,22 +498,23 @@ func CheckAllOneColumns(args ...expression.Expression) error {
...
+	v, ok := e.(*Row)
+	if ok {
+		// TODO: add check, row constructor must have >= 2 columns
+		return len(v.Values)
+	}
+
+	return 1
+}
+
+func hasSameColumn(e expression.Expression, args ...expression.Expression) error {
```
Comment (qiuyesuifeng) [inline expression/expressions/helper.go]: seems hasSameColumn only check column number length, how about rename to `hasSameColumnLen`?

File: expression/expressions/in.go [hunk 2]
```diff
@@ -158,14 +172,18 @@ func (n *PatternIn) Eval(ctx context.Context, args map[interface{}]interface{})
 			return nil, err
 		}
 
-		if g, e := len(r.GetFields()), 1; g != e {
+		if g, e := len(r.GetFields()), columnNumber(n.Expr); g != e {
 			return false, errors.Errorf("IN (%s): mismatched field count, have %d, need %d", n.Sel, g, e)
 		}
 
-		res = make([]expression.Expression, 0)
+		res = make([]interface{}, 0)
```
Comment (qiuyesuifeng) [inline expression/expressions/in.go]: how about directly use `res = []interface{}`?

File: expression/expressions/in.go [hunk 3]
```diff
@@ -158,14 +172,18 @@ func (n *PatternIn) Eval(ctx context.Context, args map[interface{}]interface{})
...
 		// evaluate select and save its result for later in expression check
 		err = r.Do(ctx, func(id interface{}, data []interface{}) (more bool, err error) {
-			res = append(res, Value{data[0]})
+			if len(data) == 1 {
+				res = append(res, data[0])
+			} else {
+				res = append(res, data)
+			}
```
Comment (qiuyesuifeng) [inline expression/expressions/in.go]: directly use `res = append(res, data...)`?

**Comment by coocood** [comment]
I suggest rename `columnNumber` to `columnCount`,  `hasSameColumn` to `hasSameColumnCount`.

File: expression/expressions/in.go (hunk 3 above)
Comment (siddontang) [inline expression/expressions/in.go]: not the same. if data has one column, the result is []interface{}, if more column, the result is [][]interface{}

**Comment by siddontang** [comment]
PTAL

File: expression/expressions/in.go (hunk 3 above)
Comment (qiuyesuifeng) [inline expression/expressions/in.go]: gotcha

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**